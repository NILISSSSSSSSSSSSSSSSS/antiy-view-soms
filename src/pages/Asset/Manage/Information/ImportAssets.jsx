import React, { Component } from 'react'
import { Button, Input, Form, message, Spin, Select } from 'antd'
import api from '@/services/api'
import { connect } from 'dva'
import PropTypes from 'prop-types'
import './style.less'

const typeRequest = [
  api.assetImportComputer,
  api.assetImportNet,
  api.assetImportSafety,
  api.assetImportStorage,
  api.assetImportOthers
]

const FormItem = Form.Item
const Option = Select.Option
//导入资产
@Form.create()
@connect(({ asset }) => ({
  loading: asset.importAssetLoading
}))
class ImportAssets extends Component{
  state = {
    uploadFile: ''
  }

  static defaultProps ={
    PrefixCls: 'Information'
  }

  static propTypes={
    hardCategoryModelNode: PropTypes.object,
    PrefixCls: PropTypes.string
  }

  submitPost = (e) => {
    e.preventDefault()
    this.props.form.validateFields((err, values)=>{
      if(err)return void(0)
      this.submitPostCB(values)
    })
  }

  // 验证文件格式
  validator = (rule, value, callback) => {
    if(value){
      const arr = value.split('.')
      const regex = /^(xls(x)?)$/ig
      const data = arr[arr.length - 1].toLowerCase()

      if(!regex.test(data)){
        callback('请上传xls或者xlsx文件')
      }
    }
    callback()
  }

  inputCB=()=>{
    const { PrefixCls } = this.props
    const files = document.getElementById(`${PrefixCls}-uploads`)
    const name = files.files[0].name
    const fileName = name.substring(0, name.lastIndexOf('.'))
    if(fileName.length > 200){
      message.info('文件名长度最多200字符，请检查')
      return void(0)
    }
    this.props.form.setFieldsValue({ file: files.value })
    this.setState({ uploadFile: files.files[0] })
  }

  submitPostCbFin=()=>{
    const { form: { resetFields }, dispatch } = this.props
    dispatch({ type: 'asset/importAssetLoading', payload: false })
    resetFields()
    this.setState({ uploadFile: '' })
  }

  submitPostCbFn=res=>{
    if(res.body.indexOf('导入失败') !== -1){
      message.warning(res.body)
    }else {
      message.success(res.body)
      this.props.handleReset()
    }
  }

  // 副作用
  submitPostCB=values=>{
    const formData = new FormData()
    formData.append('file', this.state.uploadFile)
    this.props.dispatch({ type: 'asset/importAssetLoading', payload: true })
    typeRequest[values.category - 1]( formData).then((res)=>{
      if(res && res.head && res.head.code === '200' ){
        this.submitPostCbFn(res)
      }
    }).finally(()=>{
      this.submitPostCbFin()
    })
  }

  render () {
    const { form: { getFieldDecorator }, hardCategoryModelNode, PrefixCls, loading } = this.props
    const search = {
      labelCol: {
        span: 8
      },
      wrapperCol: {
        span: 16
      }
    }
    return(
      <Spin spinning={loading}>
        <div className={PrefixCls}>
          <Form layout="horizontal" onSubmit={this.submitPost}>
            <FormItem label='选择导入类型' {...search}>
              {getFieldDecorator('category',
                { rules: [{ required: true, message: '选择导入品类' }]
                })(
                <Select allowClear placeholder="请选择"
                  getPopupContainer={triggerNode => triggerNode.parentNode}
                  showSearch={false} >
                  {hardCategoryModelNode && hardCategoryModelNode.map((item, i)=>(
                    <Option value={i + 1} key={i}>{item}</Option>)
                  )}
                </Select>
              )}
            </FormItem>
            <div>
              <FormItem {...search} label="选择文件">
                {getFieldDecorator('file', {
                  rules: [
                    { required: true, message: '选择导入文件' },
                    { validator: this.validator }]
                })(
                  <Input autoComplete="off" style={{ width: 240 }} placeholder="请选择文件"/>
                ) }
              </FormItem>
              <input
                type='file' id={`${PrefixCls}-uploads`}
                accept=".xlsx,.xls"
                key={Math.random()} //此处key是强制让元素刷新， 否则上传同一个文件时 change 不生效
                onChange={this.inputCB}/>
            </div>
            <Button style={{ visibility: 'hidden' }} htmlType='submit' id='Information-alert-post' />
          </Form>
        </div>
      </Spin>
    )
  }
}

export default ImportAssets
