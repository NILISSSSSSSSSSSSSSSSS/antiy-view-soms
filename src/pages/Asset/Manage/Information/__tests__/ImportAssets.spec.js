import React from 'react'
import { shallow } from 'enzyme'
import  ImportAssets from '../ImportAssets'
import PropTypes from 'prop-types'
// import api from '../../../../../services/api'
// jest.mock('../../../../../services/api')

// const api = {}
// const resultPro = ()=>Promise.resolve({
//   head: { code: '200' },
//   body: { }
// })
// api.assetImportComputer = jest.fn()
// api.assetImportNet = jest.fn()
// api.assetImportSafety = jest.fn()
// api.assetImportStorage = jest.fn()
// api.assetImportOthers = jest.fn()

// api.assetImportComputer.mockReturnValue(resultPro())
// api.assetImportNet.mockReturnValue(resultPro())
// api.assetImportSafety.mockReturnValue(resultPro())
// api.assetImportStorage.mockReturnValue(resultPro())
// api.assetImportOthers.mockReturnValue(resultPro())

const props = {
  hardCategoryModelNode: ['计算设备', '网络设备', '安全设备', '存储设备', '其它设备']
}

const store = {
  subscribe: () => ({
  }),
  dispatch: () =>({}),
  getState: () => ({
    asset: { importAssetLoading: false }
  })
}
const options = {
  context: { store },
  childContextTypes: { store: PropTypes.object.isRequired }
}

describe('<ImportAssets />', () => {

  const wrapper = shallow(<ImportAssets {...props}/>).dive(options).dive(options)
  const instance = wrapper.instance()

  //   console.log(instance)
  it('render <ImportAssets /> component', () => {
    expect(wrapper.find('.Information')).toHaveLength(1)
  })

  it('testFunc submitPost', () => {
    const spyFunc = jest.spyOn(instance, 'submitPost')

    const cc = {
      preventDefault: jest.fn()
    }
    spyFunc(cc)
    expect(spyFunc).toHaveBeenCalled()
  })

  it('testFunc validator', () => {
    const spyFunc = jest.spyOn(instance, 'validator')
    spyFunc(null, '12.jpg', jest.fn())
    expect(spyFunc).toHaveBeenCalled()
  })

  it('testFunc inputCB', () => {
    const spyFunc = jest.spyOn(instance, 'inputCB')
    spyFunc()
    expect(spyFunc).toHaveBeenCalled()
  })

  it('testFunc submitPostCB', () => {
    const spyFunc = jest.spyOn(instance, 'submitPostCB')
    const values = { category: 2 }
    spyFunc(values)
    expect(spyFunc).toHaveBeenCalled()
  })

  it('testFunc submitPostCbFn', () => {
    const spyFunc = jest.spyOn(instance, 'submitPostCbFn')
    const res = { body: '2' }
    spyFunc(res)
    const res2 = { body: '导入失败' }
    spyFunc(res2)
    expect(spyFunc).toHaveBeenCalled()
  })

  it('testFunc submitPostCbFin', () => {
    const spyFunc = jest.spyOn(instance, 'submitPostCbFin')
    spyFunc()
    expect(spyFunc).toHaveBeenCalled()
  })

  it('shoulder render a ImportAssets',  ()=> {
    const app = wrapper.debug()
    expect(app).toMatchSnapshot()
  })
})