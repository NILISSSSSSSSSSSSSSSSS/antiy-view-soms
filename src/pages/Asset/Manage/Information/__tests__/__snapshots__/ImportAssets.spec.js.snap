// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ImportAssets /> shoulder render a ImportAssets 1`] = `
"<Spin spinning={false} size=\\"default\\" wrapperClassName=\\"\\">
  <div className=\\"Information\\">
    <Form layout=\\"horizontal\\" onSubmit={[Function: mockConstructor]} colon={true} hideRequiredMark={false}>
      <FormItem label=\\"选择导入类型\\" labelCol={{...}} wrapperCol={{...}} hasFeedback={false}>
        <Select allowClear={true} placeholder=\\"请选择\\" getPopupContainer={[Function: getPopupContainer]} showSearch={false} transitionName=\\"slide-up\\" choiceTransitionName=\\"zoom\\" value={[undefined]} id=\\"category\\" onChange={[Function: bound bound onCollectValidate]} data-__meta={{...}} data-__field={{...}}>
          <Option value={1}>
            计算设备
          </Option>
          <Option value={2}>
            网络设备
          </Option>
          <Option value={3}>
            安全设备
          </Option>
          <Option value={4}>
            存储设备
          </Option>
          <Option value={5}>
            其它设备
          </Option>
        </Select>
      </FormItem>
      <div>
        <FormItem labelCol={{...}} wrapperCol={{...}} label=\\"选择文件\\" hasFeedback={false}>
          <Input autoComplete=\\"off\\" style={{...}} placeholder=\\"请选择文件\\" type=\\"text\\" disabled={false} value={[undefined]} id=\\"file\\" onChange={[Function: bound bound onCollectValidate]} data-__meta={{...}} data-__field={{...}} />
        </FormItem>
        <input type=\\"file\\" id=\\"Information-uploads\\" accept=\\".xlsx,.xls\\" onChange={[Function: mockConstructor]} />
      </div>
      <Button style={{...}} htmlType=\\"submit\\" id=\\"Information-alert-post\\" loading={false} ghost={false} block={false} />
    </Form>
  </div>
</Spin>"
`;
